Hello, my name is Margarita Iankova. I'm going to review your project!

You can find my comments in <span style="color:green">green</span>,<span style="color:blue">blue</span> or <span style="color:red">red</span> boxes like this:

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> if everything is done succesfully
</div>

<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
<b>Remarks:</b> if I can give some recommendations
</div>

<div class="alert alert-danger" style="box-shadow: 4px 4px 4px">
<b>Needs fixing:</b> if the block requires some corrections. Work can't be accepted with the red comments.
</div>


# General feedback

Thank you for sending your project. You've done a really good job on it! 

I would like to praise you for the structured code and very detailed comments on each step. It turned out really good research

There are also some comments for improvement. There are very few of them, you will fix them quickly. Good luck! 🍀

# General feedback №2

Thanks for sending in your project with corrections. I'm happy to say, that now your project has been accepted! Keep it up!


  
# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

## Import Pandas library of functions.

***

In [51]:
# Importing pandas libarary.
import pandas as pd

# Importing NLTK library
import nltk

# Import SnowballStemmer from nltk.
from nltk.stem import SnowballStemmer

#Setting stemmer to english.
english_stemmer = SnowballStemmer('english')

## Open the data file and have a look at the general information. 

In [52]:
# Utilizing try-except mathod in order to be able to load the 'credit_scoring_eng.csv' on the platform,
# as well as on my local computer.

try:
    #saving the data provided in the .csv in the variable credit_scoring(locally)
    credit_scoring = pd.read_csv('credit_scoring_eng.csv')
    
except:
    
    #saving the data provided in the .csv in the variable credit_scoring(platform)
    credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
    
#printing the first 10 rows of the table to take a look at the data.
credit_scoring.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

In [53]:
# Making a copy of the data so it could be manipulated without affecting the original variable, 
# in case something will go wrong.
credit_scoring_copy = credit_scoring.copy(deep=True)

<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
<b>Remarks:</b> I advise you to read the materials on this link <a href="https://www.geeksforgeeks.org/difference-between-shallow-copy-vs-deep-copy-in-pandas-dataframes/#:~:text=A%20deep%20copy%20of%20a,objects%20found%20in%20the%20original.">Shallow copy VS Deep copy</a><br>
    
In this case, if you change new DataFrame credit_scoring_copy the old one credit_scoring will also be changed
</div>

#### First we'll take a look at the data.

In [54]:
# Let's first take a closer look at the whole data set and look for any missing values as well as the data types.
credit_scoring_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-block alert-info">
As we can see, the 'days_employed' column and the 'total_income' column are both missing values.
</div>

In [55]:
# We can now use describe() to get even more details about our data.
credit_scoring_copy.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

<div class="alert alert-block alert-info">

 - That's interesting, we have a value of -1 in the 'children' column, and a maximum value of 20(:flushed:), which is not realistic.
 - Along with those negative values we have seen previously, there is a very high value at 'days_employed'.
</div>

In [56]:
# Using describe() method to check for all 'object' types.
credit_scoring_copy.describe(include='object')

education family_status gender income_type           purpose
count                 21525         21525  21525       21525             21525
unique                   15             5      3           8                38
top     secondary education       married      F    employee  wedding ceremony
freq                  13750         12380  14236       11119               797

### Conclusion

<div class="alert alert-block alert-info">

Upon first glance of the data, we can see that days_employed has some strange negative values,<br>
how can someone work negative amount of time?<br>
To better understand this, we need to take a closer look at the data.<br>
A number of the values are also unrealistically high. (Ex. row number 4), is it a typo?<br>
Is there a procedure in place when putting that data into the bank's system?<br> 
Maybe it represents something internal,<br>
or is it some sort of technical malfunction?
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Agree, good points!
</div>

<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
<b>Remarks:</b> What about general information? Step 3.1.1 logically should be here, this is not critical, but it is better to show information about the data at the very first step
</div>

<div class="alert alert-block alert-info">

 I would first try to contact whoever provided the data and ask them <br>
    About the missing income data at the total_income column. (As well as days_employed to make sense of it.)
    <br>
    However, that's not an option at the moment.<br>
 Let's see what we can do with those missing values.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Well done, you are carefully looking at the data! 
    <br> It is a good idea to write to the person responsible for the conscience of the dataset, but what if this is not possible? 
    <br>Unfortunately, this happens in real life😥
</div>

***

## Data Pre-Processing

### Processing missing values

#### Let's look at 'days_employed' column.

In [57]:
# Checking how many negative values there are in 'days_employed'.
sum(credit_scoring_copy['days_employed'] < 0)

15906

<div class="alert alert-block alert-info">

15906 values are negative in this column.<br>
    Let's take a look at the precantage.
</div>

In [58]:
# Checking the precentage of negative values in 'days_employed'.
sum(credit_scoring_copy['days_employed'] < 0) / len(credit_scoring_copy)

0.7389547038327526

<div class="alert alert-block alert-info">

 - Almost 74% of all values are negative.
</div>

In [59]:
# determinig how many missing values there are in 'days_employed'.
credit_scoring_copy['days_employed'].isnull().value_counts()

False    19351
True      2174
Name: days_employed, dtype: int64

In [60]:
# checking the precentage of missing values from total values.
credit_scoring_copy['total_income'].isnull().value_counts() / len(credit_scoring_copy)

False    0.899001
True     0.100999
Name: total_income, dtype: float64

<div class="alert alert-block alert-info">

2174 values are missing in 'days_employed' column.<br>
Which are 10% of total values.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> I like your approach to handling gaps - it's useful to know the percentage of valid and invalid values
</div>

<div class="alert alert-block alert-info">

Combining missing values and negative values in 'days_employed', we have<br>
84% of values that are not useable at their current state.<br>
    We'll need to change the negative numbers to their positive state.
    
</div>

In [61]:
# Changing values that are negative to absolute value in 'days_employed'.
credit_scoring_copy['days_employed'] = credit_scoring_copy['days_employed'].abs()

# Utilizing describe once again now that 'days_employed' has no negative values.
credit_scoring_copy.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   66914.728907     43.293380      0.817236   
std        1.381587  139030.880527     12.574584      0.548138   
min       -1.000000      24.141633      0.000000      0.000000   
25%        0.000000     927.009265     33.000000      1.000000   
50%        0.000000    2194.220567     42.000000      1.000000   
75%        1.000000    5537.882441     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

<div class="alert alert-block alert-info">

Here are a few things to consider:<br>
    - The mean is so high that it is impossible to work that many days in a lifetime, indicating that there are some fairly large outliers that push it towards the top.<br>
    - A second evidence for it is the 50% line or the median, which is a normal and reasonable value.<br>
    - In addition, if we take a look at the 'std' line (standard division), we could see that the value is excessively high, which means we have values scattered all over the place.<br>
    - And finally, when we look at 'max', we see '401755', which is impossible no matter what, it means that someone has worked non-stop for 1,100 years! (I would rather not be that person).<br>
    The following things come to mind when thinking about this,<br>
    - Maybe 'days_employed' does not describe days at all.<br>
    - Maybe the data got mixed up with another?<br>
    - Maybe the data was corrupted somewhere during the process?<br>
    
</div>

<div class="alert alert-block alert-info">
Okay, if I take into account the questions that need to be answered,<br>
As far as days_employed is concerned, I'm inclined to ignore it,<br>
Since it doesn't contain anything of use to me now but may be in the future,<br>
Therefore, I will not drop it, just ignore it for now.<br>
</div>

<div class="alert alert-danger" style="box-shadow: 4px 4px 4px">
<s><b>Needs fixing:</b> Hint: Negative values ​​are allowed. It is enough to apply the abs() method to them. But this is not the only problem with this column, I advise you to check the outliers<s>
</div>
    
<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success #2:</b> Now everything is ok 👌
</div>

 <div class="alert alert-block alert-info">
However, the 'total_income' column is important to me and to answering the questions, so let's dig in.
</div>

#### Let's look at 'total_income' columns.

In [62]:
# determinig how many missing values there are in 'total_income'.
credit_scoring_copy['total_income'].isnull().value_counts()

False    19351
True      2174
Name: total_income, dtype: int64

<div class="alert alert-block alert-info">

There are 2174 missing values in the total_income column, interesting.. it's the same amount as in 'days_employed' <br>
 - Are they conected?
</div>

In [63]:
# checking the precentage of missing values from total values.
credit_scoring_copy['total_income'].isnull().value_counts() / len(credit_scoring_copy)

False    0.899001
True     0.100999
Name: total_income, dtype: float64

<div class="alert alert-block alert-info">

 - We have about 10% of missing values in 'total_income'.
 - Because it is a quantetive value, it would be wise to fill it with the median() or the mean() of the column,<br>
    but not before we check the connection between 'days_employed' and 'total_income'.
</div>

In [64]:
# Checking for the missing values in 'total_income' and saving into 'connected_missing_values' variable.
connected_missing_values = credit_scoring_copy[credit_scoring_copy['total_income'].isnull()]

In [65]:
# Checking that all the mising values in 'days_employed' apear in the table.
connected_missing_values['days_employed'].isnull().count()

2174

<div class="alert alert-block alert-info">

In fact, all the missing values in this table are in both columns at the same row.<Br>
There may be a link between the missing values in the 'total_income' column and the 'days_employed' column.<Br>
I should investigate, however, the data is so weird at the 'days_employed' column, and it has nothing to do with the information I'm looking for,<Br> 
so I'll ignore it for now.<Br> 
In a real life situation, I would ask those who supplied the data where all the values came from.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Good reasoning 👍
</div>

#### Let's take a look at the 'children' column.

In [66]:
# Looking at 'children' column to see the different values and counts.
credit_scoring_copy['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [67]:
# Checking to see the rows with 'children' = 20.
credit_scoring_copy[credit_scoring_copy['children'] == 20].head(10)

children  days_employed  dob_years            education  education_id  \
606         20     880.221113         21  secondary education             1   
720         20     855.595512         44  secondary education             1   
1074        20    3310.411598         56  secondary education             1   
2510        20    2714.161249         59    bachelor's degree             0   
2941        20    2161.591519          0  secondary education             1   
3302        20            NaN         35  secondary education             1   
3396        20            NaN         56    bachelor's degree             0   
3671        20     913.161503         23  secondary education             1   
3697        20    2907.910616         40  secondary education             1   
3735        20     805.044438         26    bachelor's degree             0   

          family_status  family_status_id gender    income_type  debt  \
606             married                 0      M       business     0   
720             married                 0      F       business     0   
1074            married                 0      F       employee     1   
2510    widow / widower                 2      F       employee     0   
2941            married                 0      F       employee     0   
3302          unmarried                 4      F  civil servant     0   
3396            married                 0      F       business     0   
3671          unmarried                 4      F       employee     0   
3697  civil partnership                 1      M       employee     0   
3735          unmarried                 4      M       employee     0   

      total_income                                   purpose  
606      23253.578                     purchase of the house  
720      18079.798                           buy real estate  
1074     36722.966                      getting an education  
2510     42315.974  transactions with commercial real estate  
2941     31958.391                              to buy a car  
3302           NaN                         profile education  
3396           NaN                      university education  
3671     16200.879                  buying a second-hand car  
3697     18460.911                  buying a second-hand car  
3735     21952.103                        housing renovation

<div class="alert alert-block alert-info">
There are two things that going on here:<br>
    - there are 76 rows of 20 children, which is non-realistic and my guess would be a typo of 2.<br>
    - there are 47 rows of -1 children, which is non-realistic and my guess would be a typo of 1.<br>
    these errors don't count for much of the data so I feel comterable to make the changes.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Agree
</div>

***

#### Dealing with problematic columns.

<div class="alert alert-block alert-info">
    Let's start with the 'total_income' and 'children' columns.<br>
    The first column is 'total_income':<br>
    - We know there are missing values, we know it's a quantitative variable.<br>
    - We will use the mean() or the median() to fill in the data, <br>
    - It would be wise to determine if there are any outliers that can influence the data.<br>
    - It would also be wise to do it by 'gender' since men and women are paid differently, unfortunately.
</div>

#### Mean() by 'gender'.

In [68]:
# Checking the .mean() of 'total_income' by 'gender'.
credit_scoring_copy.groupby('gender')['total_income'].mean()

gender
F      24655.604757
M      30907.144369
XNA    32624.825000
Name: total_income, dtype: float64

#### Median()) by 'gender'.

In [69]:
# Checking the .median() of 'total_income' by 'gender'.
credit_scoring_copy.groupby('gender')['total_income'].median()

gender
F      21464.845
M      26834.295
XNA    32624.825
Name: total_income, dtype: float64

<div class="alert alert-block alert-info">

Due to the higher mean than median, it appears that there are some outliers that pull the data up.<br>
As the median is more representative of the data than the mean, I will fill in the missing values with the median.<br>
Let's look at the row with 'XNA' as the gender.
</div>

#### Let's look at the one row with 'XNA' as 'gender'.

In [70]:
# Taking a look at the row with 'gender' = XNA
credit_scoring_copy[credit_scoring['gender'] == 'XNA']

children  days_employed  dob_years     education  education_id  \
10701         0    2358.600502         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701     32624.825  buy real estate

<div class="alert alert-block alert-info">

Due to the fact that it is just one row, I will allow myself to change the gender to M because the total income is closer to the mean() and to the median() of M.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Analyzing total_income separately by gender is a very sensible approach.
I'm impressed with your method for changing the XNA value!
</div>

#### Changing 'XNA' in 'gender' to 'M'

In [71]:
# Changing 'XNA' to 'M'
credit_scoring_copy.loc[10701] = credit_scoring_copy.loc[10701].replace('XNA', "M")

In [72]:
# Checking that the change happend.
credit_scoring_copy['gender'].value_counts()

F    14236
M     7289
Name: gender, dtype: int64

#### Assigning variables with median by gender to fill the missing data in 'total_income'.

In [73]:
# Assigning the median by gender to the variable 'total_income_median'.
total_income_median = credit_scoring_copy.groupby('gender')['total_income'].median()

In [74]:
# Assigning the median by 'F' to the variable 'total_income_median_f'.
total_income_median_f = total_income_median.iloc[0]

# Checking to verify the assignment.
total_income_median_f

21464.845

In [75]:
# Assigning the median by 'M' to the variable 'total_income_median_m'.
total_income_median_m = total_income_median.iloc[1]

# Checking to verify the assignment.
total_income_median_m

26838.552000000003

#### Filling the missing values in 'total_income' by 'gender'.

In [76]:
# Filling missing values with gender 'F' with the median of gender 'F'.
credit_scoring_copy.loc[credit_scoring_copy['gender'] == 'F', ['total_income']] = credit_scoring_copy.loc[credit_scoring_copy['gender'] == 'F', ['total_income']].fillna(total_income_median_f)


# checking to see if there are more missing values with the gender 'F'.
credit_scoring_copy.loc[credit_scoring_copy['gender'] == 'F', ['total_income']].isnull().sum()

total_income    0
dtype: int64

In [77]:
# Filling missing values with gender F with the median of gender 'M'.
credit_scoring_copy.loc[credit_scoring_copy['gender'] == 'M', ['total_income']] = credit_scoring_copy.loc[credit_scoring_copy['gender'] == 'M', ['total_income']].fillna(total_income_median_m)


# checking to see if there are more missing values with the gender 'M'.
credit_scoring_copy.loc[credit_scoring_copy['gender'] == 'M', ['total_income']].isnull().sum()

total_income    0
dtype: int64

<div class="alert alert-block alert-info">

No more missing values in 'total_income'.<br>
Missing values have been filled by 'gender' with the median of that gender.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Good job! Indeed, in this case, it is more correct to fill in the gaps with the median rather than the average
</div>

#### Changing values in 'children' column..

In [78]:
# Changing rows with 20 children to 2.
credit_scoring_copy.loc[credit_scoring_copy['children'] == 20] = credit_scoring_copy.loc[credit_scoring_copy['children'] == 20].replace(20, 2)

#Checking the change
credit_scoring_copy['children'].value_counts()

 0    14149
 1     4818
 2     2131
 3      330
-1       47
 4       41
 5        9
Name: children, dtype: int64

In [79]:
# Changing rows with -1 children to 1.
credit_scoring_copy.loc[credit_scoring_copy['children'] == -1] = credit_scoring_copy.loc[credit_scoring_copy['children'] == -1].replace(-1, 1)

#Checking the change
credit_scoring_copy['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-block alert-info">
All the values in 'children' are now normal.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> 👍
</div>

#### Filling missing values for 'days_employed'

<div class="alert alert-block alert-info">
Because of the outliers that pull the data towards the top, I'll fill the missing values with the median of 'days_employed' as it is a reasonable number.<br>
    These missing values only account for 10% of the data so there won't be much of effect to it.
</div>

In [80]:
# Assigning the median of 'days_employed' into the variale 'days_employed_median'.
days_employed_median = credit_scoring_copy['days_employed'].median()

# Filling missing values in 'days_employed' with the median of 'days_employed'.
credit_scoring_copy['days_employed'] = credit_scoring_copy['days_employed'].fillna(days_employed_median)


# checking to see if there are more missing values in 'days_employed'.
credit_scoring_copy['days_employed'].isnull().sum()

0

<div class="alert alert-block alert-info">
No missing values at 'days_employed'.
</div>

### Conclusion

<div class="alert alert-block alert-info">

In the 'days_employed' and 'total_income' columns, the missing values appear on the same rows.<br>
It appears that there is a problem with the data of these loan seekers,<br>
It could be a problem with exporting, an issue with entering the data,<br>
It is impossible to know without further investigation at this point.<br>
Also, there were some values in 'children' that didn't make much sense, and my best guess is that they are typos.<br>
As a result of addressing the weird values within the 'children' column, as well as the missing values within <br>'total_income', The data can be investigated further.
</div>

 <div class="alert alert-danger" style="box-shadow: 4px 4px 4px">
<s><b>Needs fixing:</b> What about days_employed column? All the nan valuues should be processed here
</div> 

    
<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success #2:</b> Well done!
</div>

***

### Data type replacement

<div class="alert alert-block alert-info">

- For 'total_income' it would be better to change the data type from float to integer.<br>
With such big numbers, there is little meaning in the numbers after the decimal point.<br>
Working with it as an integer would be easier.<br>
- Also, I would change the type of the debt column to 'bool' since there are only 2 options in this column.<br>
- Finally, Having the data in 'days_employed' as a flot has very little meaning, so we better change it to int so it would be easier to work with (and on the eyes).<br>
- For the other columns, I don't see a reason to change their data type, at least for now.<br>
</div>

In [81]:
# Changing data type from float to integer for the 'total_income' column.
# Changing data type from int to bool for the 'debt' column.
# Changing data type from float to integer for the 'days_employed' column.
credit_scoring_copy = credit_scoring_copy.astype({'total_income':'int' , 'debt':'bool', 'days_employed':'int'})
credit_scoring_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null bool
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: bool(1), int64(6), object(5)
memory usage: 1.8+ MB


<div class="alert alert-block alert-info">

All types are changed to their correct state.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Well done!
</div>

### Conclusion

<div class="alert alert-block alert-info">
The only columns that needed changing were the 'total_income' column, 'days_employed' and 'debt' column.<br>
I choose the as_type() method because the values were already numbers so there was<br>
no need to convert them to numeric and than to the right type.
</div>

 <div class="alert alert-danger" style="box-shadow: 4px 4px 4px">
<s><b>Needs fixing:</b> there is one more column, the values ​​of which should be changed to the int type. But first fill in the missing values in it) <s>
</div> 
    
    
<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Agree 👍👍👍
</div>

***

### Processing duplicates

#### Looking for duplicates

<div class="alert alert-block alert-info">
In terms of duplicates in this data, I want to look for entire rows that are duplicated.<br>
It would not mean much if there is a duplicate in a single column as this data describes people<br>
It wouldn't mean much if 2 people are earning the same salary, but if the whole row is exactly the same<br>
it means we are taking this person into account twice.
</div>

In [82]:
# Checking for the amount of duplicated rows.
credit_scoring_copy.duplicated().sum()

54

<div class="alert alert-block alert-info">
There are 54 dupliactes in the data.
</div>

In [83]:
# taking a look at these row.
credit_scoring_copy[credit_scoring_copy.duplicated()].head(10)

children  days_employed  dob_years            education  education_id  \
2849          0           2194         41  secondary education             1   
4182          1           2194         34    BACHELOR'S DEGREE             0   
4851          0           2194         60  secondary education             1   
5557          0           2194         58  secondary education             1   
7808          0           2194         57  secondary education             1   
8583          0           2194         58    bachelor's degree             0   
9238          2           2194         34  secondary education             1   
9528          0           2194         66  secondary education             1   
9627          0           2194         56  secondary education             1   
10462         0           2194         62  secondary education             1   

           family_status  family_status_id gender income_type   debt  \
2849             married                 0      F    employee  False   
4182   civil partnership                 1      F    employee  False   
4851   civil partnership                 1      F     retiree  False   
5557   civil partnership                 1      F     retiree  False   
7808   civil partnership                 1      F     retiree  False   
8583           unmarried                 4      F     retiree  False   
9238             married                 0      F    employee  False   
9528     widow / widower                 2      F     retiree  False   
9627             married                 0      F     retiree  False   
10462            married                 0      F     retiree  False   

       total_income                              purpose  
2849          21464  purchase of the house for my family  
4182          21464                     wedding ceremony  
4851          21464                     wedding ceremony  
5557          21464                    to have a wedding  
7808          21464                     having a wedding  
8583          21464              supplementary education  
9238          21464      buying property for renting out  
9528          21464     transactions with my real estate  
9627          21464     transactions with my real estate  
10462         21464           buy commercial real estate

<div class="alert alert-block alert-info">
At row [4182], the 'education' column has a capitalized value.<br>
I think it would be a smart move to change all string(object) columns to lower or upper case according to their needs<br>
Then they would be taken into account if there are more duplicates in the data.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Good idea
</div>

#### Changing each column t to lower or upper case words

In [84]:
# Changing each column that has an obeject type to lower or upper case words, to make it easier to find duplicates.

# Changing 'education' to lower case.
credit_scoring_copy['education'] = credit_scoring_copy['education'].str.lower()

# Changing 'family_status' to lower case.
credit_scoring_copy['family_status'] = credit_scoring_copy['family_status'].str.lower()

# Changing 'gender' to upper case.
credit_scoring_copy['gender'] = credit_scoring_copy['gender'].str.upper()

# Changing 'income_type' to lower case.
credit_scoring_copy['income_type'] = credit_scoring_copy['income_type'].str.lower()

# Changing 'purpose' to lower case.
credit_scoring_copy['purpose'] = credit_scoring_copy['purpose'].str.lower()

In [85]:
# Checking for the amount of duplicated rows.
credit_scoring_copy.duplicated().sum()

71

<div class="alert alert-block alert-info">
Now we found 71 duplicates in the data. :)
</div>

#### Dropping duplicates

In [86]:
# Dropping the duplicates, and reseting the index of the table.
credit_scoring_copy = credit_scoring_copy.drop_duplicates().reset_index(drop=True)

In [87]:
# Checking if there are any other duplicates
credit_scoring_copy.duplicated().sum()

0

<div class="alert alert-block alert-info">
No duplicates.
</div>

### Conclusion

<div class="alert alert-block alert-info">
We have examined the data to see whether any of the rows are duplicates,<br>
to avoid counting the same person twice.<br>
Our search for duplicates turned up 54, when we looked at the data closely<br>
it turned out that some values were in uppercase and others in lowercase in 'education'.<br>
I decided to deal with that by making each column that has a string in it lower or upper case according to what it needs.<br>
71 duplicate rows were found when we looked for them again, which indicates that the process we used helped<br>
to gain a better understanding of the data.<br>
When we found the duplicates, we dropped them, so now there aren't any.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Excellent! Converting to the same case helped to detect all duplicates
</div>

***

### Categorizing Data

<div class="alert alert-block alert-info">
As I considered the best way to categorize the data, I looked at the questions I wish to answer<br> and decided it would be most helpful if I use the following categories:<br>
    - Parent and Not-Parent (first question)<br>
    - Married and Not-Married (second question)<br>
    - Level of income (third question)<br>
    - The purpose category of the loan (fourth question).
</div>

#### Categorazing by Parents and Not-Parents

In [88]:
# a function that takes the number of children and returns if a parent or not.
def is_parent(value):
    if value == 0:
        return 'Not-Parent'
    return 'Parent'

# Applying is_parent function to the 'children' column.
credit_scoring_copy['is_parent'] = credit_scoring_copy['children'].apply(is_parent)

#looking at the number of rows for 'is_parent' column.
credit_scoring_copy['is_parent'].value_counts()

Not-Parent    14091
Parent         7363
Name: is_parent, dtype: int64

<div class="alert alert-block alert-info">
I find it interesting that almost twice as many people who seek loans don't have children as those who do.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> 😮 
</div>

#### Categorazing by Married and not married

In [89]:
# a function that takes the 'family_status_id' and returns if married or not.

def is_married(value):
    if value == 0:
        return 'Married'
    return 'Not-Married'

# Applying is_married function to the 'children' column.
credit_scoring_copy['is_married'] = credit_scoring_copy['family_status_id'].apply(is_married)

#looking at the number of rows for 'is_married' column.
credit_scoring_copy['is_married'].value_counts()

Married        12339
Not-Married     9115
Name: is_married, dtype: int64

<div class="alert alert-block alert-info">
Most of the people who seek a loan are married.
</div>

#### Categorazing by Income bases

In [90]:
# a function that takes 'total_income' and returns the level of income.

def income_level(value):
    if value < 10000:
        return 'Low Income'
    elif value < 20000:
        return 'Below Average Income'
    elif value < 30000:
        return 'Average Income'
    elif value < 40000:
        return 'Higher Than Average Income'
    elif value > 40000:
        return 'High Income'

# Applying income_level function to the 'total_income' column.    
credit_scoring_copy['income_level'] = credit_scoring_copy['total_income'].apply(income_level)

#looking at the number of rows for each 'income_level'.
credit_scoring_copy['income_level'].value_counts()

Average Income                8166
Below Average Income          6443
Higher Than Average Income    3107
High Income                   2812
Low Income                     926
Name: income_level, dtype: int64

<div class="alert alert-block alert-info">
Most of the loan seekers come from the average to below average income bracket,<br> each of them being larger than the others combined.
</div>

#### Categorazing by Purpose (stemming) 

In [91]:
# Checking to see the different values in 'purpose'.
credit_scoring_copy['purpose'].value_counts()

wedding ceremony                            791
having a wedding                            768
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
housing                                     646
purchase of the house                       646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             621
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

In [92]:
# Creating a list of words that apear in all of the values to be stemmed.
words = ["wedding", "estate", "housing", "house", "property", "car", "university", "education"]
for word in words:
    print(
        'Source word - {}, after stemming - {}'.format(
            word, english_stemmer.stem(word)
        )
    )

Source word - wedding, after stemming - wed
Source word - estate, after stemming - estat
Source word - housing, after stemming - hous
Source word - house, after stemming - hous
Source word - property, after stemming - properti
Source word - car, after stemming - car
Source word - university, after stemming - univers
Source word - education, after stemming - educ


In [93]:
# a function to determine the purpose category of the loan, by stemming the 'purpose' string from the data.
def purpose_category(queries):
    for query in queries.split(" "):
        stemmed_word = english_stemmer.stem(query)
        if stemmed_word == 'wed':
            return 'Wedding'
        elif stemmed_word == 'estat':
            return 'Housing'
        elif stemmed_word == 'hous':
            return 'Housing'
        elif stemmed_word == 'properti':
            return 'Housing'
        elif stemmed_word == 'car':
            return 'Car'
        elif stemmed_word == 'univers':
            return 'Education'
        elif stemmed_word == 'educ':
            return 'Education'

# Creating a new column 'purpose_category' in the data.
credit_scoring_copy['purpose_category'] = credit_scoring_copy['purpose'].apply(purpose_category)

#looking at the number of rows for each 'purpose_category'.
credit_scoring_copy['purpose_category'].value_counts()

Housing      10811
Car           4306
Education     4013
Wedding       2324
Name: purpose_category, dtype: int64

<div class="alert alert-block alert-info">
It's interesting to observe that at first glance, one would think that 'Weddings' were the most frequent reason for borrowing money,<br> but a closer look at the data reveals that 'Housing' is the most frequent category and equal to all the others combined.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> unfortunately, many people cannot afford to buy a house without a loan 😔 
</div>

### Conclusion

<div class="alert alert-block alert-info">
The categorization of the data has inspired a hypothesis for me at this point.<br>
    - Due to the fact that most of the loan seekers are <br>
    - married without children,<br> 
    - have an average or below average income<br>
    And seeking a loan that is connected with housing. <br>
    I tend to think that these loan applicants are young couples looking to start a family,<br>
    They are planning to buy a house after getting married, but they have not yet had children,<br>
    but that is something that I would need to look into and it is out of the scope of the task.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> I really enjoy reading your conclusions) 
</div>

***

## Answer these questions

### Is there a relation between having kids and repaying a loan on time?

In [94]:
# Looking at the data, grouped by is that person is a parent and if that person is in debt.
credit_scoring_copy.groupby('is_parent')['debt'].sum().sort_values(ascending=False)

is_parent
Not-Parent    1063.0
Parent         678.0
Name: debt, dtype: float64

In [95]:
# Looking at the data, grouped by number of 'children' and if that person is in debt.
credit_scoring_copy.groupby('children')['debt'].sum().sort_values(ascending=False)

children
0    1063.0
1     445.0
2     202.0
3      27.0
4       4.0
5       0.0
Name: debt, dtype: float64

### Conclusion

<div class="alert alert-block alert-info">
By looking at the data, grouped according to if that person is a parent and if they are in debt,<br>
According to the data, people that have children pay their dues on time more often.<br>
We can also devise that people that have more kids are less likley to default on a loan, but do they take the same amount of loans?<br>
Being a parent puts a lot of pressure on you to provide for your children,<br>
You are responsible for your child.
</div>

<div class="alert alert-warning" style="box-shadow: 4px 4px 4px">
<b>Remarks:</b> Here you can also look at the relationship between the number of children and repaying a loan
</div>

***

### Is there a relation between marital status and repaying a loan on time?

In [96]:
# Looking at the data, grouped by is that person is married and if that person is in debt.
credit_scoring_copy.groupby('is_married')['debt'].sum().sort_values(ascending=False)

is_married
Married        931.0
Not-Married    810.0
Name: debt, dtype: float64

### Conclusion

<div class="alert alert-block alert-info">
Analyzing the data, grouped by whether a person is married and if a person is in debt,<br>
People who are married are more likely to default on a loan, that is interesting<br>
because you would expect them to pay on time because of their financial power compared to singles,<br>
widows ,widowers or divorced.<br>
Perhaps there is a connection between the income level of that couple, the purpose of the loan, or even their age.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Yes, most likely, marital status is not the only feature that affects loan default.
</div>

### Is there a relation between income level and repaying a loan on time?

In [97]:
# Looking at the data, grouped by income level of that person and if that person is in debt.
credit_scoring_copy.groupby('income_level')['debt'].sum().sort_values(ascending=False)

income_level
Average Income                697.0
Below Average Income          550.0
Higher Than Average Income    242.0
High Income                   194.0
Low Income                     58.0
Name: debt, dtype: float64

### Conclusion

<div class="alert alert-block alert-info">
If we look at the data grouped by income level of the person and if that person has debt,<br>
It is interesting to see that people with low income levels are more likely to pay on time.<br>
One would expect higher-income people to pay their dues on time because they have more disposable income,<br>
However, the data indicates otherwise.<br>
The purpose of the loan or perhaps the fact that psychologically the more money you earn, the higher the quality of your life,<br> 
could explain this phenomenon, because you need to maintain that expensive quality of life.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Good arguments
</div>

### How do different loan purposes affect on-time repayment of the loan?

In [98]:
# Looking at the data, grouped by 'purpose' of that loan and if that person is in debt.
# Saved into 'loan_debt_sum' variable.
loan_debt_sum = credit_scoring_copy.groupby('purpose_category')['debt'].sum().sort_values(ascending=False)
loan_debt_sum

purpose_category
Housing      782.0
Car          403.0
Education    370.0
Wedding      186.0
Name: debt, dtype: float64

In [99]:
# Looking at the data, grouped by 'purpose' and counting the number of loans in each category.
# Saved into 'loan_debt_count' variable.
loan_debt_count = credit_scoring_copy.groupby('purpose_category')['debt'].count().sort_values(ascending=False)
loan_debt_count

purpose_category
Housing      10811
Car           4306
Education     4013
Wedding       2324
Name: debt, dtype: int64

In [100]:
# Dividing 'loan_debt_count' by 'loan_debt_sum' to get the percent and saving to 'loan_debt_percent'.
loan_debt_percent = loan_debt_count / loan_debt_sum / 100
loan_debt_percent.sort_values(ascending=False)

purpose_category
Housing      0.138248
Wedding      0.124946
Education    0.108459
Car          0.106849
Name: debt, dtype: float64

### Conclusion

<div class="alert alert-block alert-info">
In looking at the data, grouped according to the purpose of the loan and if the borrower is in debt,<br>
As can be seen, the people most likely to default on a loan are those who borrow for housing,<br>
However, it is also more likely since it is the most common loan people take.<br>
It would be also interesting to see that there is a correspondents between the number of defaults and<br>
the frequency of that loan, for example the loans the are the most popular are:<br>
- Housing<br>
- Car<br>
- Education<br>
- Wedding<br>
and the numbers of defaults are descending in that exact order.
</div>

<div class="alert alert-block alert-info">
The percentage of default in each category tells a different story!<br>
- Housing still has the highest default rate, but that makes sense because it is the category with the highest number of loans.<br>
- As weddings comes second to housing in terms of defaults, maybe the bank should take caution when granting a loan for a wedding, as it's the least number of loans but with the highest amount of defaults (after housing)<br>
- In terms of percentages and total loan amounts, Education and Cars are pretty close together.
</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> Good!
    <br> Besides, it would be interesting to compare the percentage of defaults
 to the total number of loan applicants by group (optional, just for you)
</div>

## General conclusion

<div class="alert alert-block alert-info">
In spite of the fact that the data might initially appear to indicate one thing,<br>
After you have processed and cleaned the data, you come up with different conclusions.<br>
It is interesting to see and learn from this dataset, and I think I have only scratched the surface
with my investigation,<br> 
however, I do believe that there is much more that would be accounted for if a person defaults on his loan than what we are given in this dataset.<br>
There might be even more connections between the different columns of that data,<br> 
We didn't look if there is a connection between age and loan default, for example.
and we still don't understand the 'days_employed' column, and we don't know if it has some sort of impact on our findings.

</div>

<div class="alert alert-success" style="box-shadow: 4px 4px 4px">
<b>Success:</b> You have done a good research and come to interesting conclusions, keep it up!

## Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [x]  an explanation of which missing value types were detected;
- [x]  explanation for the possible causes of missing values;
- [x]  an explanation of how the blanks are filled;
- [x]  replaced the real data type with an integer;
- [x]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [x]  an explanation of which method is used to find and remove duplicates;
- [x]  description of the possible reasons for the appearance of duplicates in the data;
- [x]  data is categorized;
- [x]  an explanation of the principle of data categorization;
- [x]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [x]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [x]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [x]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [x]  conclusions are present on each stage;
- [x]  a general conclusion is made.